sklearn/metrics/_regression.py

```diff
@@ -608,13 +609,75 @@ def median_absolute_error(
     return np.average(output_errors, weights=multioutput)
 
 
+def _assemble_r2_explained_variance(
+    numerator, denominator, n_outputs, multioutput, force_finite
+):
+    """Common part used by explained variance score and :math:`R^2` score."""
+
+    if not force_finite:
+        # Standard formula, that may lead to NaN or -Inf
+        output_scores = 1 - (numerator / denominator)
+    else:
+        nonzero_denominator = denominator != 0
+        nonzero_numerator = numerator != 0
+        # Default = Zero Numerator = perfect predictions. Set to 1.0
+        # (note: even if denominator is zero, thus avoiding NaN scores)
+        output_scores = np.ones([n_outputs])
+        # Non-zero Numerator and Non-zero Denominator: use the formula
+        valid_score = nonzero_denominator & nonzero_numerator
+        output_scores[valid_score] = 1 - (
+            numerator[valid_score] / denominator[valid_score]
+        )
+        # Non-zero Numerator and Zero Denominator:
+        # arbitrary set to 0.0 to avoid -inf scores
+        output_scores[nonzero_numerator & ~nonzero_denominator] = 0.0
+
+    if isinstance(multioutput, str):
+        if multioutput == "raw_values":
+            # return scores individually
+            return output_scores
+        elif multioutput == "uniform_average":
+            # Passing None as weights to np.average results is uniform mean
+            avg_weights = None
+        elif multioutput == "variance_weighted":
+            avg_weights = denominator
+            if force_finite and not np.any(nonzero_denominator):
+                # Avoid failing on constant y or one-element arrays.
+                if not np.any(nonzero_numerator):
```

Comment:
It seems that the lines below are never executed.

Reply:
This seems like a merging issue: I remember that we moved the np.any test upwards at some point, let me check

Reply:
Indeed - they were not before the PR, too (lines 828+)

I'll add a test

Reply:
done. I changed the code actually, it was redundant